### PR TITLE
cpu/stm32f1/periph/pwm: PWM implementation for STM32F1

### PR DESCRIPTION
**Pull request by DipSwitch**
The PWM peripheral implementation for the STM32F1xx series.

### CODE REVIEW

**Comment by A-Paul** [comment]
@DipSwitch thanks for your contribution.
Looks like a smooth transition from stm32f4's pwm. I have to notice though that the coding style with preprocessor conditionals is treated as highly deprecated.
Some riot veterans may curse you and your ancestors and burn down the place your living, when they see it. :D
@haukepetersen , @kaspar030 what is your advice here?

**Comment by A-Paul** [comment]
@haukepetersen , @kaspar030 PING.

**Comment by PeterKietzmann** [comment]
@A-Paul in that case I think we could live with them. I'm just wondering about `PWM_0_PERIPH_AF`. Could you review and test anyway? Cause AFAIK we're planning a feature freeze for the next release, starting on 7th december

**Comment by DipSwitch** [comment]
@A-Paul Kill it with fire! ^_^

Should I change the driver interface? Or since this is mostly copied from F4 family maybe they can later be changed at the same time ;)

**Comment by A-Paul** [comment]
@DipSwitch I agree with  @PeterKietzmann and will start reviewing soon.
For now you haven't populated any  `periph_conf.h`, right?

**Comment by DipSwitch** [comment]
@A-Paul 

	/**
	 * @brief PWM configuration
	 * @{
	 */
	#define PWM_NUMOF           (1U)
	#define PWM_0_EN            1

	#define PWM_MAX_CHANNELS    4

	/* PWM 0 device configuration */
	#define PWM_0_DEV           TIM3
	#define PWM_0_CHANNELS      4
	#define PWM_0_CLK           (36000000U)
	#define PWM_0_CLKEN()       (RCC->APB1ENR |= RCC_APB1ENR_TIM3EN)
	#define PWM_0_CLKDIS()      (RCC->APB1ENR &= ~(RCC_APB1ENR_TIM3EN))
	/* PWM 0 pin configuration */
	#define PWM_0_PORT          GPIOB
	#define PWM_0_PORT_CLKEN()  (RCC->APB2ENR |= RCC_APB2ENR_IOPBEN)
	#define PWM_0_PIN_CH0       4
	#define PWM_0_PIN_CH1       5
	#define PWM_0_PIN_CH2       0
	#define PWM_0_PIN_CH3       1

	#define PWM_0_AFIO_MAPR_BITMASK     (0x3)
	#define PWM_0_AFIO_MAPR_OFFSET      (10)
	#define PWM_0_PERIPH_AF             (2)
	/** @} */

I tested it with the following config.

File: cpu/stm32f1/periph/pwm.c [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            pwm_clk = PWM_1_CLK;
+            PWM_1_PORT_CLKEN();
+            break;
+#endif
+    }
+
+    /* setup pins for high speed alternate function */
+    for (int i = 0; i < channels; i++) {
+        /* calculate where the MODE and CNF must be updated */
+        port->CR[pins[i] >= 8] &= ~(0xfl << (4 * (pins[i] - ((pins[i] >= 8) * 8))));
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: line 87+88:
I'm not sure if there is HW specific distinction to the f4 or why you changed the code.
It is more compact but at the same time harder to understand.

File: cpu/stm32f1/periph/pwm.c [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        port->CR[pins[i] >= 8] |= (0xbl << (4 * (pins[i] - ((pins[i] >= 8) * 8))));
+    }
+
+#if PWM_0_PERIPH_AF
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: Can you give some background?

File: cpu/stm32f1/periph/pwm.c [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    /* clear re-mapping of the peripheral and set the new re-mapping */
+    AFIO->MAPR &= ~(PWM_0_AFIO_MAPR_BITMASK << PWM_0_AFIO_MAPR_OFFSET);
+    AFIO->MAPR |= (PWM_0_PERIPH_AF << PWM_0_AFIO_MAPR_OFFSET);
+#endif
+
+#if PWM_1_PERIPH_AF
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: Forgot something here?

File: cpu/stm32f1/periph/pwm.c [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+    /* set the prescale and auto-reload register to matching values for resolution and frequency */
+    if (resolution > 0xffff || (resolution * frequency) > pwm_clk) {
+        return -2;
+    }
+    tim->PSC = (pwm_clk / (resolution * frequency)) - 1;
+    tim->ARR = resolution - 1;
+
+    /* set PWM mode */
+    switch (mode) {
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: Whole block:
You removed the conditional behaviour depending on the channel number. Why?

File: cpu/stm32f1/periph/pwm.c (hunk 1 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/pwm.c]: The F4 has a separate MODE and AFR register for the F1 it is configured via CR1 and CR2

File: cpu/stm32f1/periph/pwm.c (hunk 2 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/pwm.c]: The F1 doesn't have a pin matrix, you can only map peripherals in two or three ways.

http://www.st.com/web/en/resource/technical/document/reference_manual/CD00171190.pdf starting from page 178 you can see the remapping configuration.

File: cpu/stm32f1/periph/pwm.c [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+#endif
+    }
+
+    /* norm value to maximum possible value */
+    if (value > tim->ARR) {
+        value = (unsigned int) tim->ARR;
+    }
+
+    /* calculate the c/c offset, and set the value */
+    *((&(tim->CCR1)) + (channel * 2)) = value;
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: Not guarded against function calls with negative channel numbers.

File: cpu/stm32f1/periph/pwm.c (hunk 4 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/pwm.c]: You can't use the same timer for PWM and something else (the init will override one of both) so it doesn't hurt to configure all the CC's in the same way :)

Plus that you can have 2 two channels but want to use CC3 and CC4 which are mapped on CCMR2. If you have the conditional behavior the channels won't be configured at all.

Maybe this needs to be fixed for the F4 aswell now i'm thinking of it?

File: cpu/stm32f1/periph/pwm.c (hunk 1 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/pwm.c]: The pin configuration can be found here:
http://www.st.com/web/en/resource/technical/document/reference_manual/CD00171190.pdf on page 171

File: cpu/stm32f1/periph/pwm.c (hunk 3 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/pwm.c]: Jup, fixed it and added support for remapping of timer > 9.

File: cpu/stm32f1/periph/pwm.c (hunk 4 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/pwm.c]: Ah, I see that this is actually invalid, but you can't only set CH3 to PWM.

File: cpu/stm32f1/periph/pwm.c [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    /* norm value to maximum possible value */
+    if (value > tim->ARR) {
+        value = (unsigned int) tim->ARR;
+    }
+
+    /* set the PWM value to the channel comperator */
+    switch (channel)
+    {
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: Curly braces on same line as block expression.

**Comment by A-Paul** [comment]
@DipSwitch thanks for addressing the comments and providing some inline documentation.
(For rebase without warning I'll put a spell on you. :D)

**Comment by A-Paul** [comment]
I'm preparing on testing on a iotlab-m3. There is a very limited number of pins routed to a connector.
As I can see I should use *PB9* with *TIM4,CH4*. Can you please assist writing the config?

This is what I'm assuming for now.

```
/* PWM 0 device configuration */
#define PWM_0_DEV           TIM4
#define PWM_0_CHANNELS      4
#define PWM_0_CLK           (36000000U)
#define PWM_0_CLKEN()       (RCC->APB1ENR |= RCC_APB1ENR_TIM4EN)
#define PWM_0_CLKDIS()      (RCC->APB1ENR &= ~(RCC_APB1ENR_TIM4EN))
/* PWM 0 pin configuration */
#define PWM_0_PORT          GPIOB
#define PWM_0_PORT_CLKEN()  (RCC->APB2ENR |= RCC_APB2ENR_IOPBEN)
...
```
tbc

**Comment by DipSwitch** [comment]
Should be something like this:


	/**
	 * @brief PWM configuration
	 * @{
	 */
	#define PWM_NUMOF           (1U)
	#define PWM_0_EN            1

	#define PWM_MAX_CHANNELS    4

	/* PWM 0 device configuration */
	#define PWM_0_DEV           TIM4
	#define PWM_0_CHANNELS      4
	#define PWM_0_CLK           (36000000U)
	#define PWM_0_CLKEN()       (RCC->APB1ENR |= RCC_APB1ENR_TIM4EN)
	#define PWM_0_CLKDIS()      (RCC->APB1ENR &= ~(RCC_APB1ENR_TIM4EN))
	/* PWM 0 pin configuration */
	#define PWM_0_PORT          GPIOB
	#define PWM_0_PORT_CLKEN()  (RCC->APB2ENR |= RCC_APB2ENR_IOPBEN)
	#define PWM_0_PIN_CH0       6
	#define PWM_0_PIN_CH1       7
	#define PWM_0_PIN_CH2       8
	#define PWM_0_PIN_CH3       9
	/** @} */

**Comment by A-Paul** [comment]
If travis should compile and test your implementation at least one board has to use the new feature.
Please provide an updated `boards/iotlab-m3/Makefile.features` (add `FEATURES_PROVIDED += periph_pwm`) and `boards/iotlab-m3/include/periph_conf.h` in your PR.

I did it in my local checkout. This is what *make* spit out.

```
/home/aar680/git/RIOT/cpu/stm32f1/periph/pwm.c: In function 'pwm_init':
/home/aar680/git/RIOT/cpu/stm32f1/periph/pwm.c:93:14: error: comparison between pointer and integer [-Werror]
     if ((tim > TIM12_BASE && tim < TIM14_BASE) || (tim > TIM15 && tim < TIM11)) {
              ^
/home/aar680/git/RIOT/cpu/stm32f1/periph/pwm.c:93:34: error: comparison between pointer and integer [-Werror]
     if ((tim > TIM12_BASE && tim < TIM14_BASE) || (tim > TIM15 && tim < TIM11)) {
                                  ^
/home/aar680/git/RIOT/cpu/stm32f1/periph/pwm.c:92:20: error: variable 'MAPR' set but not used [-Werror=unused-but-set-variable]
     __IO uint32_t *MAPR = &(AFIO->MAPR);
                    ^
cc1: all warnings being treated as errors
```

**Comment by DipSwitch** [comment]
Sorry about that, was doing to many things at once...

Fixed it and added pin configuration and AF selection to the source.

File: cpu/stm32f1/periph/pwm.c [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+#include <stdint.h>
+#include <string.h>
+
+#include "cpu.h"
+#include "periph/pwm.h"
+#include "periph_conf.h"
+
+/* ignore file in case no PWM devices are defined */
+#if PWM_0_EN || PWM_1_EN || 1
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: @DipSwitch, the ` || 1` invokes a true "power mode" (file is always compiled ). But, I have the idea that two Travis fails may be caused by that.

**Comment by DipSwitch** [comment]
Fixed it :)

**Comment by DipSwitch** [comment]
Squash?

**Comment by A-Paul** [comment]
Travis shows some strange errors here. @authmillenon, @x3ro someone to tell me whats the matter?

**Comment by DipSwitch** [comment]
#4420 should fix this, I kick travis again.

**Comment by A-Paul** [comment]
@DipSwitch, tried again, but still not working. You may squash meanwhile.

**Comment by DipSwitch** [comment]
Done, maybe this triggers travis again. Last job wouldn't restart it seems.

**Comment by DipSwitch** [comment]
Travis agrees, but I guess the feature freeze kicked in. :)

**Comment by OlegHahm** [comment]
We have some delays because of Travis problems yesterday. No feature freeze yet.

**Comment by A-Paul** [comment]
I will run the test on *iotlab-m3* today. If it is successful I give my ack.

**Comment by A-Paul** [comment]
Tested on *iotlab-m3* it generaly works but duty cycle frequencies are incorrect.

|requested|pwm_out|return|
|-------------|-----------|--------|
|          100|        200|    100|
|        1000|      2000|  1000|
|        5000|    10310|  5000|

Also, I haven't noticed before that `pwm_init` returns just the requested frequency and not the frequency resulting from the prescaler settings.

**Comment by haukepetersen** [comment]
how urgent is it, to get this PR in? If not too much, I would propose to wait for #3644 and adapt it accordingly to get the driver in a 'modern' state from the start...

**Comment by DipSwitch** [comment]
Yes, it's ok to wait for the next release.

**Comment by haukepetersen** [comment]
cool, then lets try to make give these drivers a 'modern' look and with such the `stm32f1` can be used as reference on good periph drivers for others!

**Comment by A-Paul** [comment]
Hey @DipSwitch just FYI.  I assume the "Waiting For Other PR" is related to #3644. PR #3644 has been superseded by #4638, which has already been merged.

**Comment by DipSwitch** [comment]
I'll get to it in the weekend :) thx

**Comment by OlegHahm** [comment]
Which weekend? ;-)

**Comment by DipSwitch** [comment]
The weekends were I'm still working on my house. :p There was a little more
work to do then we expected ;) we are now waiting for the floor of my
mancave, then I can get my hardware out of the boxes and start testing =)

**Comment by OlegHahm** [comment]
No sweat! Just wanted to make sure that nothing gets lost here. :)

**Comment by PeterKietzmann** [comment]
I don't think this PR is waiting for others anymore? @DipSwitch can you rebase @A-Paul can you test please

**Comment by DipSwitch** [comment]
Yes, it needs a minor fix for calculating the bus clockspeed. And I need to
wait for my nucleo-f103 to arrive.

**Comment by kYc0o** [comment]
@DipSwitch ping

**Comment by DipSwitch** [comment]
Still need to be refactored. Allot, and it's time for sleep, so maybe tomorrow :)

File: cpu/stm32f1/include/stm32f103xe.h [hunk 8]
```diff
@@ -1767,6 +1767,10 @@ typedef struct
 
 #define AFIO_MAPR_TIM4_REMAP                 ((uint32_t)0x00001000)        /*!< TIM4_REMAP bit (TIM4 remapping) */
 
+/*!< TIM4_REMAP configuration */
```
Comment (A-Paul) [inline cpu/stm32f1/include/stm32f103xe.h]: + following 2 lines. Can't this be done in an other place than the original vendors file?

File: boards/iotlab-m3/include/periph_conf.h [hunk 9]
```diff
@@ -168,6 +168,21 @@ static const uart_conf_t uart_config[] = {
...
 
+/**
+ * @brief PWM configuration
+ * @{
+ */
+#define PWM_EN              1
+
+static const pwm_conf_t pwm_config[] =
+{
+    { TIM4, AFIO_MAPR_TIM4_REMAP_NOREMAP, 2, 1, {
```
Comment (A-Paul) [inline boards/iotlab-m3/include/periph_conf.h]: Magic numbers

File: cpu/stm32f1/periph/pwm.c [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    conf->dev->CR1 = 0;
+    conf->dev->CR2 = 0;
+    for (int idx = 0; idx < channels; ++idx) {
+        *ccr(dev, idx) = 0;
+    }
+
+    /* set the pre-scale and auto-reload register to matching values
+     * for resolution and frequency
+     */
+    if (res > UINT16_MAX || (res * freq) > pwm_clk) {
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: `res > UINT16_MAX` is pointless.

File: cpu/stm32f1/include/stm32f103xe.h (hunk 8 above)
Comment (DipSwitch) [inline cpu/stm32f1/include/stm32f103xe.h]: These are the only two missing... I guess I could place them somewhere else though

**Comment by DipSwitch** [comment]
Addressed, rebased etc.

This seems to work

**Comment by A-Paul** [comment]
Trying to do a hw-test with *iotlab-m3*. `make` gives me:

```
/home/aar680/git/RIOT/boards/iotlab-m3/include/periph_conf.h:179:43: error: 'RCC_APB1ENR_TIM4' undeclared  here (not in a function)
     { TIM4, AFIO_MAPR_TIM4_REMAP_NOREMAP, RCC_APB1ENR_TIM4, APB1, {
                                           ^
```

With *nucleo-f103* compiling succeeds. But I have no board to run it.

**Comment by DipSwitch** [comment]
Sorry for that, was a little to enthusiastic.

**Comment by A-Paul** [comment]
@DipSwitch , built and tested on `iotlab-m3` with *tests/periph_pwm* successful.

Please squash.

**Comment by DipSwitch** [comment]
squashed, rebased, pushed

**Comment by DipSwitch** [comment]
Fixed fox and sparc-core periph_conf.h files

**Comment by DipSwitch** [comment]
Add a fix for `tests/driver_servo`

File: cpu/stm32f1/periph/pwm.c [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    conf->dev->CR2 = 0;
+    for (int idx = 0; idx < channels; ++idx) {
+        *ccr(dev, idx) = 0;
+    }
+
+    /* set the pre-scale and auto-reload register to matching values
+     * for resolution and frequency
+     */
+    if ((res * freq) > pwm_clk) {
+        return -2;
```
Comment (A-Paul) [inline cpu/stm32f1/periph/pwm.c]: Sorry, if I haven't seen this early. This "error code" doesn't exist anymore.

**Comment by A-Paul** [comment]
There are again some macro values that may be lucky to be embraced by warming parenthesis. ;-)

**Comment by kYc0o** [comment]
Feature freeze -> postponed.

**Comment by kYc0o** [comment]
And please rebase, it would be nice to have it for the next release.

**Comment by miri64** [comment]
Postponed due to Feature Freeze

**Comment by kYc0o** [comment]
This was addressed by #6186 ?

**Comment by PeterKietzmann** [comment]
No it wasn't.  #6186 didn't cover stm32f1 based platforms. @DipSwitch can you adapt this PR according to #6186 please?! You think you can make it until end of  next week?

**Comment by kYc0o** [comment]
Needs rebase, and since it's a new feature, maybe should be postponed.

**Comment by kYc0o** [comment]
@DipSwitch do you mind to rebase this one?

**Comment by PeterKietzmann** [comment]
Implementation is outdated and no progress for the last year. Will close with memo label set.

**Pull request closed without merging**